File: taurus-protect-sdk-go/pkg/protect/helper/address_signature_verifier.go
package helper

import (
	"fmt"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/crypto"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// VerifyAddressSignature verifies the address signature using the HSM slot public key.
// The signed data is the blockchain address string. The public key is found by locating
// a user with the HSMSLOT role in the rules container.
//
// Returns nil if verification succeeds, IntegrityError otherwise.
func VerifyAddressSignature(address *model.Address, rulesContainer *model.DecodedRulesContainer) error {
	if address == nil {
		return &model.IntegrityError{Message: "address cannot be nil"}
	}
	if rulesContainer == nil {
		return &model.IntegrityError{Message: "rulesContainer cannot be nil"}
	}

	// Check for signature presence (mandatory)
	if address.Signature == "" {
		return &model.IntegrityError{
			Message: fmt.Sprintf("Address %s has no signature", address.ID),
		}
	}

	// Check for address string presence
	if address.Address == "" {
		return &model.IntegrityError{
			Message: fmt.Sprintf("Address %s has no blockchain address to verify", address.ID),
		}
	}

	// Get the HSM public key (cached in rules container)
	hsmPublicKey := rulesContainer.GetHsmPublicKey()
	if hsmPublicKey == nil {
		return &model.IntegrityError{
			Message: "No user with HSMSLOT role found in rules container",
		}
	}

	// Verify the signature
	addressData := []byte(address.Address)
	valid, err := crypto.VerifySignature(hsmPublicKey, addressData, address.Signature)
	if err != nil {
		return &model.IntegrityError{
			Message: fmt.Sprintf("Address signature verification failed for address %s: %v", address.ID, err),
		}
	}

	if !valid {
		return &model.IntegrityError{
			Message: fmt.Sprintf("Address signature verification failed for address %s", address.ID),
		}
	}

	return nil
}

// VerifyAddressSignatures verifies signatures for multiple addresses.
// Returns the first verification error encountered, or nil if all signatures are valid.
func VerifyAddressSignatures(addresses []*model.Address, rulesContainer *model.DecodedRulesContainer) error {
	for _, address := range addresses {
		if err := VerifyAddressSignature(address, rulesContainer); err != nil {
			return err
		}
	}
	return nil
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/helper/AddressSignatureVerifier.java
package com.taurushq.sdk.protect.client.helper;

import com.google.common.base.Strings;
import com.taurushq.sdk.protect.client.model.Address;
import com.taurushq.sdk.protect.client.model.IntegrityException;
import com.taurushq.sdk.protect.client.model.rulescontainer.DecodedRulesContainer;
import com.taurushq.sdk.protect.openapi.auth.CryptoTPV1;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.PublicKey;
import java.security.SignatureException;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Helper class for verifying address signatures using HSM slot public keys.
 * <p>
 * Address signatures are verified against the public key of a user
 * with the HSMSLOT role in the rules container.
 */
public final class AddressSignatureVerifier {

    private AddressSignatureVerifier() {
        // Utility class - prevent instantiation
    }

    /**
     * Verifies the address signature using the HSM slot public key from the rules container.
     * <p>
     * The signed data is the blockchain address string only. The public key is found
     * by locating a user with the HSMSLOT role in the rules container.
     *
     * @param address        the address to verify
     * @param rulesContainer the decoded rules container containing HSM public keys
     * @throws IntegrityException if the address has no signature, no user with HSMSLOT role is found,
     *                            or signature verification fails
     */
    public static void verifyAddressSignature(Address address, DecodedRulesContainer rulesContainer)
            throws IntegrityException {

        checkNotNull(address, "address cannot be null");
        checkNotNull(rulesContainer, "rulesContainer cannot be null");

        // Check for signature presence (mandatory)
        if (Strings.isNullOrEmpty(address.getSignature())) {
            throw new IntegrityException("Address " + address.getId() + " has no signature");
        }

        // Check for address string presence
        if (Strings.isNullOrEmpty(address.getAddress())) {
            throw new IntegrityException("Address " + address.getId() + " has no blockchain address to verify");
        }

        // Get the HSM public key (cached in rules container)
        PublicKey hsmPublicKey = rulesContainer.getHsmPublicKey();
        if (hsmPublicKey == null) {
            throw new IntegrityException("No user with HSMSLOT role found in rules container");
        }

        // Verify the signature
        try {
            byte[] addressData = address.getAddress().getBytes(StandardCharsets.UTF_8);
            boolean valid = CryptoTPV1.verifyBase64Signature(hsmPublicKey, addressData, address.getSignature());

            if (!valid) {
                throw new IntegrityException("Address signature verification failed for address " + address.getId());
            }
        } catch (NoSuchAlgorithmException | InvalidKeyException | SignatureException e) {
            throw new IntegrityException("Address signature verification failed for address " + address.getId(), e);
        }
    }
}
